### ISSUE TITLE
Speeding up the Travis verifications

### ISSUE THREAD

**Issue by benjub** [metamath/set.mm#1202]
I noticed that Travis verification is pretty slow.  It actually does several checks on set.mm, iset.mm, .raw.html files and others, even if those have not been modified.  But most pull requests modify only one or two of these files.  I don't know the language it's written with, but there is probably a function that checks whether a file has been modified, so that for instance, at line 56 of .travis.yml,
> metamath 'read set.mm' 'markup mmset.raw.html mmset.html /ALT /CSS' quit

could be replaced by something like
> IF modified(mmset.raw.html) THEN metamath 'read set.mm' 'markup mmset.raw.html mmset.html /ALT /CSS' quit

And similarly with the other tests, especially the lengthy ones involving mmj2.  This would probably save time and energy.  Is this doable ?

**Comment by digama0** [comment]
I'm amused that you think 3 minutes for the complete build is slow. The lean mathlib compilation takes 2 hours, and requires some hacking on travis to circumvent the time limit.

Checking all the files doesn't actually take that long, but we are taking advantage of the fact that compilation is so fast to throw every tool we have at the mm files. The longest running programs are mmj2 and checkmm, which mostly reflects inefficiency of the program itself rather than the work to be done. The easiest way to speed up the build is to just drop the verifiers that take too long, but at least for now it seems like we can afford to run a bit longer for that extra assurance.

TL;DR: Travis is plenty fast already.

**Comment by david-a-wheeler** [comment]
I suspect it could be significantly sped up by running the different tests in parallel. One way would be to move things into a makefile and using make -j. As Mario noticed, our verification process is so much faster than everyone else's that we haven't bothered.

**Comment by benjub** [comment]
I'm happy you're amused, but none of you answered my only question. Let's forget about speed, my question is simply: can we replace, roughly,
> verify set.mm
> verify iset.mm
> ...

with
> if modified(set.mm) then verify set.mm
> if modified(iset.mm) then  verify iset.mm
> ...

?  Does there exist such a function "modified" ?

Even if the total time were 3 seconds, it would still bother me to make poor Travis verify all files even when they are unchanged (and as I wrote, few pull requests modify more than one or two files, so no need to spend time and energy checking all of them). Plus, isn't git all about tracking changes in files?

**Comment by digama0** [comment]
To your concrete question, I think it could be done if you query git in the right way, but I don't think there is a simple builtin for it.

The more important point for me is what we hope to get out of Travis. If we want to make it faster, why not just use an efficient verifier? What are we gaining by using multiple verifiers? To me, the point is that Travis is there to pick up our messes. It always starts from a clean slate, and gives us feedback about the full build. The cost of this is that it takes a bit longer and has to work a little harder. Incremental builds are convenient, but they are complicated, and I have seen bugs caused from bad caching behavior before.

Incremental builds should be done on your machine, when you are sitting there waiting for it. It's just for you anyway, and the quick turnaround is more important than absolute correctness. But when you put a new commit on master, that's for everyone, and I would rather see full builds there because it's easier to get right, even if it costs a bit more, unless the resource costs become extreme (as in the mathlib example).

**Comment by benjub** [comment]
Thanks for this answer. I thought my proposal would be simple to implement, but if the test condition is harder to carry out, then the cure might be worse than the disease, indeed, and it's better to stay as is.

**Comment by david-a-wheeler** [comment]
I believe it is possible, though it is not easy and would take some research. What's more, it really isn't just changing of the mm files. For example, if the metamath source code changes, you want to use it to check all the mm files, even if the mm files haven't changed.

But if we switch your question to how can we make verification faster, I think that is definitely possible.

**Comment by david-a-wheeler** [comment]
I'll look over https://docs.travis-ci.com/user/speeding-up-the-build/ to see about some options. We could also switch from Travis, but the same issues would apply.

**Comment by benjub** [comment]
> For example, if the metamath source code changes, you want to use it to check all the mm files, even if the mm files haven't changed.

Good point.

Thanks @david-a-wheeler ; as @digama0 wrote, correctness is more important than speed, and the verification is not too long anyway.  I just thought I saw a quick and easy fix, but that's not the case, so please do not spend too much time on this.

I'm going to close the issue tomorrow unless there is some more activity.

**Comment by tirix** [comment]
Noting that David _did_ propose a way to do this:

> One way would be to move things into a makefile and using make -j.

**Comment by vvs-** [comment]
If I understand it correctly every time when Travis checks a build it has to clone a set.mm repository again? Caching build artifacts around could be a non-trivial exercise and bug prone. You should take into account not only bugs in your own scripts but in the used tools as well.

I can recollect that git is particularly poor at dealing with the file timestamps which upsets make. There is still a very old bug where git might completely ruin timestamps depending on the order of its operations, e.g `git reset; git reset --hard` gives a different result than `git reset --hard` alone.

**Comment by benjub** [comment]
We need Mario to formalize and prove git in metamath zero before we can proceed !

**Comment by david-a-wheeler** [comment]
@benjub - don't close this issue.  I think trying to "minimize work" is unnecessarily risky and in practice won't help much (most changes are to set.mm, and that requires most of the work to check).

However, if your goal is to make it *faster*, the obvious solution is to parallelize things, which I think is fairly easy.

**Comment by david-a-wheeler** [comment]
I just created pull request  https://github.com/metamath/set.mm/pull/1204 that brings verification time to around 1min 20sec, better than halving the verification time.

**Comment by david-a-wheeler** [comment]
There are other steps we can do to speed things up further.  Also, since the jobs are now in parallel, we can add additional verifiers without spending extra time (as long as they're at least as fast).

**Comment by benjub** [comment]
Nice improvement, @david-a-wheeler ! Tell me when you think I can close this issue.

**Comment by benjub** [comment]
A question: it looks like currently, the generation of the *discouraged.new files is done by mmj2.  However, metamath.c can also do it, and the associated job is probably less long ? Maybe this and the lines currently 79,80 (beginning with "echo") can also be moved to the "metamath.c" job ?

**Comment by nmegill** [comment]
@benjub The metamath.exe 'show discouraged' is currently painfully slow (almost 3 minutes) compared to mmj2 and needs to be rewritten.  It is on my to-do list.

**Comment by david-a-wheeler** [comment]
I think we can close this issue.  The verification time is half of what it used to be, about 1min 20seconds now.  That's also 1/120th of Lean's verification time.  No doubt we can keep improving, but the "easy" improvements are done.

**Comment by benjub** [comment]
I'm closing tomorrow if no further activity. Just noting that when metamath.c has a faster "show discouraged", we can add it to .travis.yml (in addition to, or maybe in place of, the mmj2 equivalent).

**Comment by david-a-wheeler** [comment]
We can certainly add "show discouraged" for metamath.c.  However, I recommend doing that in *addition* to the check done by mmj2.

One *very* unusual fact about Metamath is that there are so many verifiers.  Since we can run them in parallel, we should take advantage of them to run *many* verifiers to provide higher confidence that the databases are correct.  If the mmj2 or Metamath C programs have an error, there's still a decent chance that another tool will still catch an error.

**Comment by jkingdon** [comment]
> I'm amused that you think 3 minutes for the complete build is slow. The lean mathlib compilation takes 2 hours

Well, yes, this is one of my major reasons for bothering with metamath rather than learning a "real" prover (in addition to, well, just not having made it very far up the learning curve with coq or Isabelle, both of which I got as far as downloading and if memory serves running).

This could veer into a rather long and serious discussion, but probably most of the people who care already know about the tradeoffs involved in the design of these systems, so I'll leave it at that.
